File: HW/HW2/Main.py
# The isBadVersion API is already defined.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        start, end = 1, n
        while start < end:
            mid = start + (end - start) // 2
            if isBadVersion(mid):
                end = mid  # look on left side of mid
            else:
                start = mid + 1  # look on the right side of mid
        return start  # this point would be the first bad version

File: HW/HW2/Main.java
package HW.HW2;
/* Supposing the isBadVersion API is defined in the parent class VersionControl:
      boolean isBadVersion(int version); */
      public class Solution extends VersionControl {
        public int firstBadVersion(int n) {
            int start = 1, end = n;
            while (start < end) {
                int mid = start + (end - start) / 2;
                if (isBadVersion(mid)) {
                    end = mid; // look on left side of mid
                } else {
                    start = mid + 1; // look on the right side of mid
                }
            }
            return start; // this point would be the first bad version
        }
    }
